src/main/java/com/nimbusds/jose/crypto/impl/ECDH1PUCryptoProvider.java

```diff
@@ -0,0 +1,270 @@
+                throw new JOSEException(AlgorithmSupportMessage.unsupportedEncryptionMethod(
+                        header.getEncryptionMethod(),
+                        EncryptionMethod.Family.AES_CBC_HMAC_SHA));
+            }
+
+            if (contentEncryptionKey != null) { // Use externally supplied CEK
+                cek = contentEncryptionKey;
+            } else { // Generate the CEK according to the enc method
+                cek = ContentCryptoProvider.generateCEK(enc, getJCAContext().getSecureRandom());
+            }
+
+            JWECryptoParts encrypted = ContentCryptoProvider.encrypt(header, clearText, cek, null, getJCAContext());
+
+            SecretKey sharedKey = ECDH1PU.deriveSharedKey(header, Z, encrypted.getAuthenticationTag(), getConcatKDF());
+            encryptedKey = Base64URL.encode(AESKW.wrapCEK(cek, sharedKey, getJCAContext().getKeyEncryptionProvider()));
+
+            return new JWECryptoParts(
+                    header,
+                    encryptedKey,
+                    encrypted.getInitializationVector(),
+                    encrypted.getCipherText(),
+                    encrypted.getAuthenticationTag()
+            );
+        }
+
+        throw new JOSEException("Unexpected JWE ECDH algorithm mode: " + algMode);
+    }
+
+    /**
+     * Decrypts the encrypted JWE parts using the specified shared secret ("Z").
+     */
+    protected byte[] decryptWithZ(final JWEHeader header,
+                      final SecretKey Z,
+                      final Base64URL encryptedKey,
+                      final Base64URL iv,
+                      final Base64URL cipherText,
+                      final Base64URL authTag)
+        throws JOSEException {
+
+        final JWEAlgorithm alg = header.getAlgorithm();
+        final ECDH.AlgorithmMode algMode = ECDH1PU.resolveAlgorithmMode(alg);
+
+        // Derive shared key via concat KDF
+        getConcatKDF().getJCAContext().setProvider(getJCAContext().getMACProvider()); // update before concat
+
+        JWEHeader updatedHeader = new JWEHeader.Builder(header).
+                iv(iv).
+                authTag(authTag).
+                build();
+
+        final SecretKey cek;
+
+        if (algMode.equals(ECDH.AlgorithmMode.DIRECT)) {
+            cek = ECDH1PU.deriveSharedKey(updatedHeader, Z, getConcatKDF());
+        } else if (algMode.equals(ECDH.AlgorithmMode.KW)) {
+            if (encryptedKey == null) {
+                throw new JOSEException("Missing JWE encrypted key");
+            }
+
+            SecretKey sharedKey = ECDH1PU.deriveSharedKey(updatedHeader, Z, authTag, getConcatKDF());
+            cek = AESKW.unwrapCEK(sharedKey, encryptedKey.decode(), getJCAContext().getKeyEncryptionProvider());
+        } else {
+            throw new JOSEException("Unexpected JWE ECDH algorithm mode: " + algMode);
+        }
+
+        return ContentCryptoProvider.decrypt(header, encryptedKey, iv, cipherText, authTag, cek, getJCAContext());
```

Comment:
I agree, but it's an nimbus jose api. It exists for backward compatibility with deprecated code. I don't want to make any big changes which may have a negative impact on contribution to upstream here.

Reply:
Just pass the `null` parameter instead of the `encodedKey` one to the decrypt function. No need to keep secret keys, even encrypted, in memory longer than required.